eth/watchers/senderwatcher_test.go

```diff
@@ -0,0 +1,469 @@
+	assert.Equal(info, senderInfo)
+
+	// If map entry exists parse event log
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	expectedDeposit := senderInfo.Deposit.Add(senderInfo.Deposit, new(big.Int).SetBytes(newDepositEvent.Data))
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Zero(info.Deposit.Cmp(expectedDeposit))
+
+	// if block removed, make RPC call
+	blockEvent.Type = blockwatch.Removed
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Equal(info, lpEth.SenderInfo)
+}
+
+func TestFundReserveEvent(t *testing.T) {
+	assert := assert.New(t)
+	senderInfo := &pm.SenderInfo{
+		Deposit: big.NewInt(10),
+		Reserve: big.NewInt(5),
+	}
+	lpEth := &eth.StubClient{
+		SenderInfo: senderInfo,
+	}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	// We only keep cache for addresses that exist in the map
+	_, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+
+	header := defaultMiniHeader()
+	newReserveEvent := newStubReserveFundedLog()
+	header.Logs = append(header.Logs, newReserveEvent)
+
+	blockEvent := &blockwatch.Event{
+		Type:        blockwatch.Added,
+		BlockHeader: header,
+	}
+
+	go sw.Watch()
+	defer sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+
+	// If map entry is nil , make RPC call and init the map entry
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, ok := sw.senders[stubSender]
+	assert.True(ok)
+	assert.Equal(info, lpEth.SenderInfo)
+
+	// If map entry exists parse event log
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	expectedReserve := senderInfo.Reserve.Add(senderInfo.Reserve, new(big.Int).SetBytes(newReserveEvent.Data))
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Zero(info.Reserve.Cmp(expectedReserve))
+
+	// if block removed, make RPC call
+	blockEvent.Type = blockwatch.Removed
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Equal(info, lpEth.SenderInfo)
+}
+
+func TestWithdrawalEvent(t *testing.T) {
+	assert := assert.New(t)
+	senderInfo := &pm.SenderInfo{
+		Deposit: big.NewInt(10),
+		Reserve: big.NewInt(5),
+	}
+	lpEth := &eth.StubClient{
+		SenderInfo: senderInfo,
+	}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	header := defaultMiniHeader()
+	newWithdrawalEvent := newStubWithdrawalLog()
+	header.Logs = append(header.Logs, newWithdrawalEvent)
+
+	blockEvent := &blockwatch.Event{
+		Type:        blockwatch.Added,
+		BlockHeader: header,
+	}
+
+	go sw.Watch()
+	defer sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+
+	// If map entry is nil it is left nil since reserve and deposit are withdrawn
```

Comment:
I think a more accurate comment is probably "If map entry is nil it is left nil since we do not care about the address in the log"

Reply:
fixed in a18ffc1